### DISCUSSION TITLE
Nil concurrentDB error

### DISCUSSION THREAD

**Discussion by Oggy4th** [pocketbase/pocketbase#3755] [Q&A]
Hello, there 

When I stopped pocketbase in server below error printed in errors.log.

```logtalk
2023/11/15 07:26:16 http: panic serving xxx.180.219.230:53184: runtime error: invalid memory address or nil pointer dereference
goroutine 9332667 [running]:
net/http.(*conn).serve.func1()
	/Users/oggy/sdk/go1.21.4/src/net/http/server.go:1868 +0xb9
panic({0x1643680?, 0x2c04130?})
	/Users/oggy/sdk/go1.21.4/src/runtime/panic.go:920 +0x270
github.com/pocketbase/pocketbase/daos.(*Dao).ConcurrentDB(...)
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/base.go:70
github.com/pocketbase/pocketbase/daos.(*Dao).DB(...)
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/base.go:62
github.com/pocketbase/pocketbase/daos.(*Dao).ModelQuery(0x0, {0x20f4a18?, 0xc00e1eb2b0?})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/base.go:108 +0x40
github.com/pocketbase/pocketbase/daos.(*Dao).CollectionQuery(...)
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/collection.go:18
github.com/pocketbase/pocketbase/daos.(*Dao).FindCollectionByNameOrId(0x2c2f320?, {0xc001612940, 0xf})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/collection.go:41 +0x59
github.com/pocketbase/pocketbase/daos.(*Dao).FindRecordById(0x1754c80?, {0xc001612940?, 0x181f97c?}, {0xc001612980, 0xf}, {0x0, 0x0, 0xc0058779b0?})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/record.go:138 +0x45
github.com/pocketbase/pocketbase/daos.(*Dao).FindAuthRecordByToken(0x15a9840?, {0xc00a5ab520, 0xcf}, {0xc00065c900, 0x32})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/record.go:428 +0x13f
github.com/pocketbase/pocketbase/apis.InitApi.LoadAuthContext.func5.1({0x2100c28, 0xc000fe9a80})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/apis/middlewares.go:241 +0x1df
github.com/labstack/echo/v5/middleware.RemoveTrailingSlashConfig.ToMiddleware.func1.1({0x2100c28, 0xc000fe9a80})
	/Users/oggy/go/pkg/mod/github.com/labstack/echo/v5@v5.0.0-20230722203903-ec5b858dab61/middleware/slash.go:136 +0x206
github.com/labstack/echo/v5.(*Echo).ServeHTTP(0xc0009d0000, {0x20ea340, 0xc011cbb340}, 0xc00cbaab00)
	/Users/oggy/go/pkg/mod/github.com/labstack/echo/v5@v5.0.0-20230722203903-ec5b858dab61/echo.go:670 +0x22a
net/http.serverHandler.ServeHTTP({0xc010a0d470?}, {0x20ea340?, 0xc011cbb340?}, 0x6?)
	/Users/oggy/sdk/go1.21.4/src/net/http/server.go:2938 +0x8e
net/http.(*conn).serve(0xc001843ef0, {0x20ec6c8, 0xc000163890})
	/Users/oggy/sdk/go1.21.4/src/net/http/server.go:2009 +0x5f4
created by net/http.(*Server).Serve in goroutine 67
	/Users/oggy/sdk/go1.21.4/src/net/http/server.go:3086 +0x5cb
```

**Comment by ganigeorgiev** [comment]
This seems similar to your other reports in https://github.com/pocketbase/pocketbase/discussions/3166, https://github.com/pocketbase/pocketbase/discussions/3093, https://github.com/pocketbase/pocketbase/discussions/3067, but again without details about your setup or any reproducible steps I'm not sure how to help you debug it.

As mentioned in some of the linked discussion, currently if the application is abruptly killed/terminated it is possible to see the above panic message when there is a delayed log write but it shouldn't have any real impact (_other than maybe some missing logs_) because it is a "handled/recovered" panic (_we just print the raw error message at the moment_).

The logs write currently are executed concurrently after the request has been processed to avoid blocking. The logs write operation may panic in case the app is abruptly killed/terminated because the db connections will be closed and the delayed log write is not notified about this.
This will be fixed with the upcoming logs generalization in v0.20.0 where instead of firing a new goroutine for each log write (_we still can have only 1 writer at a time_), we batch the logs in memory and bulk insert them after some threshold is reached (or when the logs api endpoints are invoked _and/or right before app termination_).

**Reply by ganigeorgiev** [reply]
Another possibility for the panic based on the stack trace could be if some request(s) were already in processing when the app was terminated, but again the above shouldn't happen in normal app operation and you can safely ignore the error.

**Reply by ganigeorgiev** [reply]
@Oggy4th I saw the GitHub invite, but please try to create a **minimal public** reproducible repo.

But again you can safely ignore the error message since most likely it is caused either by a delayed log write or already started request processing.

I'll consider with the next release either to move the app resources clean up after all hooks and/or at least provide a more helpful message.

**Reply by Oggy4th** [reply]
Hello, there

I create repo and invited you. You can find my code from there.

There have below 2 errors. So errors collecting and so everything is not responding until restart pocketbase.

Yesterday evening pocketbase suddenly stopped to respond and continuously loading everything. When I stop almost 40000 lines errors printed in errors.log.

```logtalk
2023/11/14 17:34:51 RECOVERED FROM PANIC: runtime error: invalid memory address or nil pointer dereference
2023/11/14 17:34:51 goroutine 13761322 [running]:
runtime/debug.Stack()
	/Users/oggy/sdk/go1.21.4/src/runtime/debug/stack.go:24 +0x5e
github.com/pocketbase/pocketbase/tools/routine.FireAndForget.func1.1()
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/tools/routine/routine.go:26 +0xaa
panic({0x1643680?, 0x2c04130?})
	/Users/oggy/sdk/go1.21.4/src/runtime/panic.go:914 +0x21f
github.com/pocketbase/pocketbase/daos.(*Dao).lockRetry(0x2c03ed0?, 0x20dadc0?)
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/base.go:378 +0x3f
github.com/pocketbase/pocketbase/daos.(*Dao).Save(0x1632a00?, {0x20f49a0, 0xc04b6e1c70})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/base.go:263 +0x6c
github.com/pocketbase/pocketbase/daos.(*Dao).SaveRequest(0x0?, 0x0?)
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/request.go:69 +0x1d
github.com/pocketbase/pocketbase/apis.bindRealtimeApi.ActivityLogger.func1.1.1()
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/apis/middlewares.go:343 +0x4e
github.com/pocketbase/pocketbase/tools/routine.FireAndForget.func1()
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/tools/routine/routine.go:30 +0x87
created by github.com/pocketbase/pocketbase/tools/routine.FireAndForget in goroutine 616423
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/tools/routine/routine.go:18 +0xa5

2023/11/14 17:34:52 http: panic serving 202.180.219.24:64611: runtime error: invalid memory address or nil pointer dereference
goroutine 13760816 [running]:
net/http.(*conn).serve.func1()
	/Users/oggy/sdk/go1.21.4/src/net/http/server.go:1868 +0xb9
panic({0x1643680?, 0x2c04130?})
	/Users/oggy/sdk/go1.21.4/src/runtime/panic.go:920 +0x270
github.com/pocketbase/pocketbase/daos.(*Dao).ConcurrentDB(...)
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/base.go:70
github.com/pocketbase/pocketbase/daos.(*Dao).DB(...)
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/base.go:62
github.com/pocketbase/pocketbase/daos.(*Dao).ModelQuery(0x0, {0x20f4a18?, 0xc04bc4ea90?})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/base.go:108 +0x40
github.com/pocketbase/pocketbase/daos.(*Dao).CollectionQuery(...)
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/collection.go:18
github.com/pocketbase/pocketbase/daos.(*Dao).FindCollectionByNameOrId(0x2c2f320?, {0xc04be1cde0, 0xf})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/collection.go:41 +0x59
github.com/pocketbase/pocketbase/daos.(*Dao).FindRecordById(0x1754c80?, {0xc04be1cde0?, 0x181f97c?}, {0xc04be1ce20, 0xf}, {0x0, 0x0, 0xc0102859b0?})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/record.go:138 +0x45
github.com/pocketbase/pocketbase/daos.(*Dao).FindAuthRecordByToken(0x15a9840?, {0xc048b6c1c7, 0xcf}, {0xc0054aa800, 0x32})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/record.go:428 +0x13f
github.com/pocketbase/pocketbase/apis.InitApi.LoadAuthContext.func5.1({0x2100c28, 0xc006697680})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/apis/middlewares.go:241 +0x1df
github.com/labstack/echo/v5/middleware.RemoveTrailingSlashConfig.ToMiddleware.func1.1({0x2100c28, 0xc006697680})
	/Users/oggy/go/pkg/mod/github.com/labstack/echo/v5@v5.0.0-20230722203903-ec5b858dab61/middleware/slash.go:136 +0x206
github.com/labstack/echo/v5.(*Echo).ServeHTTP(0xc0006f6b40, {0x20ea340, 0xc04bc2aa80}, 0xc04be58000)
	/Users/oggy/go/pkg/mod/github.com/labstack/echo/v5@v5.0.0-20230722203903-ec5b858dab61/echo.go:670 +0x22a
net/http.serverHandler.ServeHTTP({0x20e0c50?}, {0x20ea340?, 0xc04bc2aa80?}, 0x6?)
	/Users/oggy/sdk/go1.21.4/src/net/http/server.go:2938 +0x8e
net/http.(*conn).serve(0xc048444090, {0x20ec6c8, 0xc00058cbd0})
	/Users/oggy/sdk/go1.21.4/src/net/http/server.go:2009 +0x5f4
created by net/http.(*Server).Serve in goroutine 52
	/Users/oggy/sdk/go1.21.4/src/net/http/server.go:3086 +0x5cb

```

**Reply by ganigeorgiev** [reply]
I don't think I'll be able to help you and I think we are going in circles.

As mentioned in the linked discussions, the above is a "recovered" panic and it shouldn't result in unresponsiveness.

Also note that if the app process was abruptly terminated or crashed for some reason it won't restart automatically on its own unless there is a systemd service to keep it up.

As mentioned earlier, the above is expected if the app is in the process of being terminated and there is a background db task running concurrently.

I'll try with the next release to detect this case and eventually provide a more helpful error message but again it shouldn't be the cause of your app not being responsive.

**Reply by Oggy4th** [reply]
I wrote a log in SaveRequest()
```go
// SaveRequest upserts the provided Request model.
func (dao *Dao) SaveRequest(request *models.Request) error {
	if dao == nil {
		fmt.Println("Dao is nil in SaveRequest()")
		return errors.New("Dao is nil in SaveRequest()")
	}

	return dao.Save(request)
}
```
This is the error log
```logtalk
2023/11/15 11:16:02 http: panic serving 202.200.219.50:49931: runtime error: invalid memory address or nil pointer dereference
goroutine 1958077 [running]:
net/http.(*conn).serve.func1()
	/Users/oggy/sdk/go1.21.4/src/net/http/server.go:1868 +0xb9
panic({0x1643680?, 0x2c04130?})
	/Users/oggy/sdk/go1.21.4/src/runtime/panic.go:920 +0x270
github.com/pocketbase/pocketbase/daos.(*Dao).ConcurrentDB(...)
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/base.go:70
github.com/pocketbase/pocketbase/daos.(*Dao).DB(...)
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/base.go:62
github.com/pocketbase/pocketbase/daos.(*Dao).ModelQuery(0x0, {0x20f4b40?, 0xc001f03110?})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/base.go:108 +0x40
github.com/pocketbase/pocketbase/daos.(*Dao).CollectionQuery(...)
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/collection.go:18
github.com/pocketbase/pocketbase/daos.(*Dao).FindCollectionByNameOrId(0x2c2f320?, {0xc00125fd10, 0xf})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/collection.go:41 +0x59
github.com/pocketbase/pocketbase/daos.(*Dao).FindRecordById(0x1754c80?, {0xc00125fd10?, 0x181f97c?}, {0xc00125fd50, 0xf}, {0x0, 0x0, 0xc0014619b0?})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/record.go:138 +0x45
github.com/pocketbase/pocketbase/daos.(*Dao).FindAuthRecordByToken(0x15a9840?, {0xc0021a0340, 0xcf}, {0xc000b26100, 0x32})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/record.go:428 +0x13f
github.com/pocketbase/pocketbase/apis.InitApi.LoadAuthContext.func5.1({0x2100dc8, 0xc002a5a100})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/apis/middlewares.go:241 +0x1df
github.com/labstack/echo/v5/middleware.RemoveTrailingSlashConfig.ToMiddleware.func1.1({0x2100dc8, 0xc002a5a100})
	/Users/oggy/go/pkg/mod/github.com/labstack/echo/v5@v5.0.0-20230722203903-ec5b858dab61/middleware/slash.go:136 +0x206
github.com/labstack/echo/v5.(*Echo).ServeHTTP(0xc000758000, {0x20ea4e0, 0xc002b609a0}, 0xc002de2700)
	/Users/oggy/go/pkg/mod/github.com/labstack/echo/v5@v5.0.0-20230722203903-ec5b858dab61/echo.go:670 +0x22a
net/http.serverHandler.ServeHTTP({0xc001c16750?}, {0x20ea4e0?, 0xc002b609a0?}, 0x6?)
	/Users/oggy/sdk/go1.21.4/src/net/http/server.go:2938 +0x8e
net/http.(*conn).serve(0xc000cb3320, {0x20ec868, 0xc000878330})
	/Users/oggy/sdk/go1.21.4/src/net/http/server.go:2009 +0x5f4
created by net/http.(*Server).Serve in goroutine 36
	/Users/oggy/sdk/go1.21.4/src/net/http/server.go:3086 +0x5cb
**Dao is nil in SaveRequest()**
```
This is my [digitalocean metric](https://github.com/pocketbase/pocketbase/assets/44497354/6178e616-8b2f-4c21-8a12-d3e97bb2344e) when it shuts down

Is it possible that panic logs built up to the point it makes the server unresponsive? Or is it that I have too many realtime connections but I increased LimitNOFILE in systemd
```
[Unit]
Description = pocketbase

[Service]
Type           = simple
User           = root
Group          = root
LimitNOFILE    = 131072
Restart        = always
RestartSec     = 5s
StandardOutput = append:/root/pocketbase/errors.log
StandardError  = append:/root/pocketbase/errors.log
ExecStart      = /root/pocketbase/pocketbase serve --http="example.com:80" --https="example.com:443"

[Install]
WantedBy = multi-user.target
```

**Reply by ganigeorgiev** [reply]
> Is it possible that panic logs built up to the point it makes the server unresponsive?

No, I doubt that this is the case.

I will try one last time but I really don't know how to explain it better.

The above panic is expected to be triggered **only when the application is in the process of being terminated** (aka. the cleanup procedure was invoked) and there is a db operation running concurrently in the background.

**In other words it shouldn't be the reason why your app is being unresponsive or crashing. It is a result of it.** Something else is causing your application to be unresponsive but I don't know what exactly. If you suspect that the activity logs are the culprit, you can disable them entirely from the _Admin UI > Settings > Application > "Logs max days retention" input_ and set it to `0`.

As mentioned earlier, with the upcoming logs generalization, the logs write performance and its CPU usage will be improved in v0.20.0 release because they will be written in the database on bulks (_currently although each log write is executed in its own goroutine concurrently, the write queries are still sequential and can be processed only one by one_).

**Reply by Oggy4th** [reply]
Okay, thank you very much for your response. I'll set it to 0 for now and see it crashes again also the weird part is it freezes so systemd can't restart it automatically I have to manually start the service everytime. If setting it to 0 fixes the issue I'll let you know.

**Reply by jaximo98** [reply]
I'm currently testing how Pocketbase and extending it with Go works. But I'm experiencing the "same" issue. The only code I have is a simple main, copy-pasted from the docs and a single query.


Notes:
1. Fresh project, no db files created.
2. I ran the code once, made a admin account, and then added a single collection called "players" with 1 column "name"
3. Wrote the query and on execution it fails. No matter what query I send. 
4. I'm exiting the process by doing CTRL+C whenever I want to stop the server. 

```go
package main

import (...)

func main() {
    app := pocketbase.New()

    // serves static files from the provided public dir (if exists)
    app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
        e.Router.GET("/*", apis.StaticDirectoryHandler(os.DirFS("./pb_public"), false))
        return nil
    })
   
   res, err := app.Dao().DB().
	NewQuery("ALTER TABLE players ADD alive BOOL DEFAULT TRUE").
	Execute()
   fmt.Println(res, err)

    if err := app.Start(); err != nil {
        log.Fatal(err)
    }
}
```


```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x1cb7e3b]

goroutine 1 [running]:
github.com/pocketbase/pocketbase/daos.(*Dao).ConcurrentDB(...)
        C:/Users/jax/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/base.go:70
github.com/pocketbase/pocketbase/daos.(*Dao).DB(...)
        C:/Users/jax/go/pkg/mod/github.com/pocketbase/pocketbase@v0.19.4/daos/base.go:62
main.main()
        C:/Users/jax/GolandProjects/Abyssmmo/cmd/server/main.go:38 +0x9b
```

Are there any extra considerations I should have? From your answer I understand that the query hits the db, something goes wrong and it crashes. But I do not understand what is causing it. Is there any way I can debug the cause of it?

**Reply by ganigeorgiev** [reply]
@Jaximo98-dev Your issue is not related to the above discussion. `app.Dao().DB()` is not initialized yet. You need to wrap it in a [`OnAfterBootstrap` hook](https://pocketbase.io/docs/go-event-hooks/#onafterbootstrap) (or move the query inside the `OnBeforeServe` handler).

_Additionally, I don't know what is your use case, but I'd advice against running direct schema change queries unless you really want that. If you want to update a collection programatically you can check [Collection operations](https://pocketbase.io/docs/go-collections/) and [Migrations](https://pocketbase.io/docs/go-migrations/) guides._

**Reply by jaximo98** [reply]
Thanks that worked! Also, you were right, running direct schema change queries has some issues. For instance the dashboard would not show the new tables (even though I can confirm that they were created properly). Doing it via collection operations worked like a charm, I don't know how I missed the tab in the docs... 

One question not related to this issue but small enough to let it slip, is PB docs page in any of the repositories under pocketbase? I couldn't find it either, I'm so sorry.

**Reply by ganigeorgiev** [reply]
> is PB docs page in any of the repositories under pocketbase?

The PocketBase docs source can be found in https://github.com/pocketbase/site.
